File: new_system/gcs/ReelController/ReelController.cpp
// JDW 2016-2-10
// Copyright Creare 2016
#include "ReelController.hpp"
namespace gcs {
    const unsigned int ReelController::QC_PER_TURN = 1024*4;
    const unsigned int ReelController::MOTOR_MAX_RPM = 10000;
    const unsigned int ReelController::GEARBOX_MAX_INPUT_RPM = 8000;

    ReelController::ReelController(std::string port, double reel_diam_cm) :
        motor_controller(port), reel_diameter_m(reel_diam_cm / 100.0), gear_ratio(26),
        profile_accel_mpss(0.6), profile_decel_mpss(0.6), last_commanded_tether_length_m(nan(""))
    {
        std::cout << "Initializing with port: " << port << " reel diameter:" 
            << reel_diameter_m << "m and qc/rev: " << QC_PER_TURN << std::endl;
        motor_controller.open();
        init();
    }

    long ReelController::motorPositionFromTetherLength(double tether_length_m) {
        return QC_PER_TURN * (-tether_length_m / (M_PI * reel_diameter_m));
    }

    double ReelController::tetherLengthFromMotorPosition(int motor_position) {
        return (-(double)motor_position / QC_PER_TURN) * (M_PI * reel_diameter_m);
    }

    unsigned int ReelController::motorRpmFromTetherMps(double tether_mps) {
        // Do not invert sign, since the motor controller only refers to rates in the positive
        return (60.0 * tether_mps / (M_PI * reel_diameter_m));
    }

    double ReelController::tetherMpsFromMotorRpm(unsigned int motor_rpm) {
        // Do not invert sign, since the motor controller only refers to rates in the positive
        return (double)motor_rpm * (M_PI * reel_diameter_m) / 60.0;
    }

    void ReelController::init() {
        if(motor_controller.isOpen()) {
            // Note: By convention, we will not apply settings to the motor controller here.
            // Instead, we will use the EPOS Studio to set up settings.
            // This is because we have a working set of settings, and because EPOS Studio provides
            // a more informative explanation of a lot of the settings, so it's harder to lose track
            // of, for example, which sensor you're configuring.
            
            // Get and remember gear ratio
            int num = motor_controller.getGearRatioNumerator();
            int den = motor_controller.getGearRatioDenominator();
            gear_ratio = (double)num / (double)den;
            std::cout << "Got gear ratio - " << num << ":" << den << " = " << gear_ratio << std::endl;
            
            // Get and remember acceleration/deceleration
            unsigned int tether_rpm       ;
            unsigned int tether_accel_rpms;
            unsigned int tether_decel_rpms;
            motor_controller.getPositionProfile(&tether_rpm, &tether_accel_rpms, &tether_decel_rpms);
            profile_accel_mpss = tetherMpsFromMotorRpm(tether_accel_rpms);
            profile_decel_mpss = tetherMpsFromMotorRpm(tether_decel_rpms);
                
            // Set up
            motor_controller.setOperatingMode(EposMotorController::EPOS_OPMODE_PROFILE_POSITION_MODE);
            motor_controller.clearFaultAndEnable();
        } else {
            std::cout << "Motor controller not open." << std::endl;
        }
    }

    // Passthrough functions
    void ReelController::haltMovement() { motor_controller.haltMovement(); }
    void ReelController::disable() { motor_controller.disable(); }
    void ReelController::clearFaultAndEnable() { motor_controller.clearFaultAndEnable(); }
    bool ReelController::isEnabled() { return motor_controller.isEnabled(); }

    void ReelController::setTetherToHome() {
        last_commanded_tether_length_m = 0;
    }

    void ReelController::setTetherLength(double desired_length_m) {
        // TBD: add offsets
        last_commanded_tether_length_m = desired_length_m;
        double desired_motor_position = motorPositionFromTetherLength(desired_length_m);
        std::cout << "Moving motor to " << desired_motor_position << std::endl;
        motor_controller.moveToPosition((long)desired_motor_position);
    }

    double ReelController::setMaxTetherSpeed(double max_tether_mps) {
        unsigned int max_payout_rpm = motorRpmFromTetherMps(max_tether_mps);
        if(max_payout_rpm * gear_ratio > MOTOR_MAX_RPM)
        { std::cout << "Limiting max RPM to motor max" << std::endl; max_payout_rpm = MOTOR_MAX_RPM / gear_ratio; }
        if(max_payout_rpm * gear_ratio > GEARBOX_MAX_INPUT_RPM)
        { std::cout << "Limiting max RPM to gearbox max" << std::endl;  max_payout_rpm = GEARBOX_MAX_INPUT_RPM / gear_ratio; }
        std::cout << "Setting max payout velocity to " << max_tether_mps << "mps = " << max_payout_rpm << "RPM." << std::endl;
        motor_controller.setMaxVelocity(max_payout_rpm);
        return max_payout_rpm;
    }

    double ReelController::getMaxTetherSpeed() {
        double max_payout_rpm = motor_controller.getMaxVelocity();
        return tetherMpsFromMotorRpm(max_payout_rpm);
    }

    void ReelController::setTetherSpeed(double tether_mps) {
        unsigned int tether_rpm        = motorRpmFromTetherMps(fabs(tether_mps));
        unsigned int tether_accel_rpms = motorRpmFromTetherMps(profile_accel_mpss);
        unsigned int tether_decel_rpms = motorRpmFromTetherMps(profile_decel_mpss);
        std::cout << "Setting profile to /" << tether_accel_rpms
                  << " " << tether_rpm << " \\" << tether_decel_rpms << std::endl;
        motor_controller.setPositionProfile(tether_rpm, tether_accel_rpms, tether_decel_rpms);
        if(!isnan(last_commanded_tether_length_m)) {
            setTetherLength(last_commanded_tether_length_m);
        }
    }

    double ReelController::getTetherSpeed() {
        unsigned int tether_rpm       ;
        unsigned int tether_accel_rpms;
        unsigned int tether_decel_rpms;
        motor_controller.getPositionProfile(&tether_rpm, &tether_accel_rpms, &tether_decel_rpms);
        
        return tetherMpsFromMotorRpm(tether_rpm);
    }

    void ReelController::setTetherAccelDecel(double accel_mpss, double decel_mpss)
    {
        // Store these for later; we'll send them every time we set the speed
        profile_accel_mpss = fabs(accel_mpss);
        profile_decel_mpss = fabs(decel_mpss);
    }

    double ReelController::getTetherLength() {
        int cur_motor_position = motor_controller.getPosition();
        std::cout << "Motor position=" << cur_motor_position << " ";
        return tetherLengthFromMotorPosition(cur_motor_position);
    }
}

File: new_system/gcs/ReelController/ReelController.py
"""
Reel Control top object

Object hierarchy as follows (connections represent composition not inheritance):

ReelController (in ReelController.py)
  | 
PyMotorController (in PyMotorController.pyx) - Note: this is added to Python's global import list by running "python setup.py build_ext --inplace"
  |
EposMotorController (in EposMotorController.hpp/cpp)

"""

import math
import logging

class ReelController:
    def __init__(self, interface='USB0', reel_diam_m=0.127):
        # Motor settings
        self._QC_PER_TURN = 1024*4;
        self._MOTOR_MAX_RPM = 10000;
        self._GEARBOX_MAX_INPUT_RPM = 8000;
        
        # Sensor settings
        self._N_PER_ADC_COUNT        = 0.0045203
        self._SENSOR_BASELINE_COUNTS = 7848
        self._SENSOR_DEADBAND_COUNTS = 100
        self._T_DEADBAND_N           = self._N_PER_ADC_COUNT * self._SENSOR_DEADBAND_COUNTS

        # Reel system settings
        self._reel_diam_m            = reel_diam_m # Need to set this first so the conversion methods to work
        self._MAX_RPM                = 60
        self._MIN_RPM                = 6
        self._MAX_MPS                = self.tetherMpsFromReelRpm(self._MAX_RPM)
        self._MIN_MPS                = self.tetherMpsFromReelRpm(self._MIN_RPM)
        self._L_MAX_SPEED_M          = 10 # length no longer limits reel speed beyond this range
        self._T_MAX_SPEED_N          = 5  # After this many newtons of force, don't limit payout rate
        self._KT_MPS_PER_N           =  self._L_MAX_SPEED_M / (self._T_MAX_SPEED_N - self._T_DEADBAND_N)
        self._KL_MPS_PER_M           = (self._MAX_MPS - self._MIN_MPS) / self._L_MAX_SPEED_M
        self._home_pos_m             = 0
        self._motor_is_halted        = False
        self._last_commanded_m       = 0
        
        # For easier testing, create mock objects if the real ones fail.
        try:
            from TensionSensor import TensionSensor
            self._tension_sensor = TensionSensor(adc_per_ct=self._N_PER_ADC_COUNT, adc_baseline=self._SENSOR_BASELINE_COUNTS)
            self._tension_sensor.readTension()
        except:
            logging.warning("Cannot connect to tension sensor!  Will be using mock sensor instead.")
            from MockTensionSensor import MockTensionSensor
            self._tension_sensor = MockTensionSensor()

        try:
            from PyMotorController import PyMotorController, SensorType
            self._mc = PyMotorController(interface)
        except:
            logging.warning("Cannot connect to motor controller!  Will be using mock motor controller instead.")
            from MockPyMotorController import MockPyMotorController
            self._mc = MockPyMotorController()
        # Configure motor controller
        self._gear_ratio = self._mc.getGearRatioNumerator() / self._mc.getGearRatioDenominator()
        self.youAreHome()
        self._mc.clearFaultAndEnable() # movement may occur after this point

    # Conversion functions
    def motorPositionFromTetherLength(self, tether_length_m):
        return self._QC_PER_TURN * (-tether_length_m / (math.pi * self._reel_diam_m))
    def tetherLengthFromMotorPosition(self, motor_position):
        return (-motor_position / self._QC_PER_TURN) * (math.pi * self._reel_diam_m);
    def tetherMpsFromReelRpm(self, reel_rpm):
        return reel_rpm   * (math.pi * self._reel_diam_m / 60)
    def reelRpmFromTetherMps(self, tether_mps):
        return tether_mps / (math.pi * self._reel_diam_m / 60)
    

    def __del__(self):
        self.stopMoving()
        del self._mc
        del self._tension_sensor
        
    def youAreHome(self):
        """ Consider the tether's current position to be 0m """
        self._home_pos_m = self.getTetherLengthM()
        self._last_commanded_m = 0
        self.update()
    
    def update(self):
        """
        Updates the maximum speed of the motor controller
        based on the tension from the tension sensor while 
        the reel spools to the correct tether length.
        Call this method frequently in your main loop.
        """
        
        current_length = self.getTetherLengthM()
        target_length  = self.getTargetTetherLengthM()
        
        # Update the maximum speed of the motor controller differently
        # if it's spooling out vs reeling in.  When spooling out, we
        # care about the tension, since we don't want to let line out until
        # the UAV will take up the slack.  When reeling in, we want the
        # UAV to slow down as it approaches the landing site.
        length_limited_speed  = self._KL_MPS_PER_M * current_length + self._MIN_MPS
        if current_length > target_length:
            # Reeling in
            logging.info("Reeling in because %fm > %fm, length limits us to %fmps"%(current_length, target_length, length_limited_speed))
            speed_limit = min(self._MAX_MPS, length_limited_speed)
        else: # Stationary OR reeling out
            # Apply tension deadband
            tension_n = self._tension_sensor.readTension()
            if tension_n < self._T_DEADBAND_N:
                logging.info("Within deadband because %fN < %fN"%(tension_n, self._T_DEADBAND_N))
                speed_limit = 0
            else:
                tension_n -= self._T_DEADBAND_N # Prevent "step" up when exiting deadband
                tension_limited_speed = self._KT_MPS_PER_N * tension_n
                logging.info("Reeling out, length limits us to %fmps, tension limits us to %fmps"%(length_limited_speed, tension_limited_speed))
                speed_limit = min(self._MAX_MPS, length_limited_speed, tension_limited_speed)
        
        # Apply speed limit
        logging.info("Changing speed limit from %fmps to %fmps"%(self.getMaxTetherSpeedMps(),speed_limit))
        if speed_limit == 0:
            #self._mc.haltMovement()
#            self._motor_is_halted = True
            speed_limit = 0.01
#        else:
#            position_to_recommand = None # If the motor was halted, need to give it the target location again
#            if self._motor_is_halted:
#                position_to_recommand = self._mc.getTargetPosition()
        actual_max_mps = self.setMaxTetherSpeedMps(speed_limit)
#            if position_to_recommand != None:
#                logging.info("Recommanding motor to position %f"%position_to_recommand)
#                self._mc.moveToPosition(position_to_recommand)
#                self._motor_is_halted = False
        logging.info("Max tether mps wound up being %f"%actual_max_mps)

    def stopMoving(self):
        self._mc.haltMovement()
    
    def setTetherLengthM(self, tether_length_m):
        self._last_commanded_m = tether_length_m
        desired_motor_position = self.motorPositionFromTetherLength(tether_length_m + self._home_pos_m)
        self._mc.moveToPosition(desired_motor_position)
        self.update()
        
    def getTetherLengthM(self):
        cur_motor_position = self._mc.getPosition()
        return self.tetherLengthFromMotorPosition(cur_motor_position) - self._home_pos_m
        
    def getTargetTetherLengthM(self):
        tgt_motor_position = self._mc.getTargetPosition()
        return self.tetherLengthFromMotorPosition(tgt_motor_position) - self._home_pos_m
        
    def setMaxTetherSpeedMps(self, max_tether_mps):
        max_payout_rpm = self.reelRpmFromTetherMps(max_tether_mps);
        if max_payout_rpm * self._gear_ratio > self._MOTOR_MAX_RPM:
            max_payout_rpm = self._MOTOR_MAX_RPM / self._gear_ratio
        if max_payout_rpm * self._gear_ratio > self._GEARBOX_MAX_INPUT_RPM:
            max_payout_rpm = self._GEARBOX_MAX_INPUT_RPM / self._gear_ratio
        logging.info("Setting reel max RPM=%f"%(max_payout_rpm))
        self._mc.setMaxVelocity(max_payout_rpm);
        return self.tetherMpsFromReelRpm(max_payout_rpm);

    def getMaxTetherSpeedMps(self):
        max_payout_rpm = self._mc.getMaxVelocity();
        return self.tetherMpsFromReelRpm(max_payout_rpm);
